node/coinstacks/common/api/src/evm/gasOracle.ts

```diff
@@ -0,0 +1,162 @@
+import { ethers } from 'ethers'
+import { Logger } from '@shapeshiftoss/logger'
+import { Fees } from './models'
+import { NewBlock } from '@shapeshiftoss/blockbook'
+
+export interface GasOracleArgs {
+  logger: Logger
+  provider: ethers.providers.JsonRpcProvider
+}
+
+interface TxFees {
+  gasPrices: Array<number>
+  maxPriorityFees: Array<number>
+}
+
+interface BlockFees extends TxFees {
+  baseFeePerGas?: number
+}
+
+type EstimatedFees = Record<string, Fees>
+
+export class GasOracle {
+  private readonly totalBlocks = 20
+  private readonly logger: Logger
+  private readonly provider: ethers.providers.JsonRpcProvider
+
+  private feesByBlock = new Map<string, BlockFees>()
+
+  constructor(args: GasOracleArgs) {
+    this.logger = args.logger.child({ namespace: ['gasOracle'] })
+    this.provider = args.provider
+  }
+
+  async start() {
+    const height = await this.provider.getBlockNumber()
+    const length = this.totalBlocks - 1
+    const startingBlock = height - length
+    const blocks = Array.from<number, string | number>({ length }, (_, index) => index + 1 + startingBlock)
+    blocks.push('pending')
+    await Promise.all(blocks.map(async (blockNumber) => this.update(blockNumber)))
+  }
+
+  getBaseFeePerGas(): string | undefined {
+    return this.feesByBlock.get('pending')?.baseFeePerGas?.toString()
+  }
+
+  // estimate fees based on the current oracle state
+  async estimateFees(percentiles: Array<number>): Promise<EstimatedFees> {
+    const { baseFeePerGas, number } = (await this.provider.send('eth_getBlockByNumber', ['pending', false])) as {
+      number: string
+      baseFeePerGas?: string
+    }
+
+    const latestBlock = Number(number) - 1
+
+    await this.validate(latestBlock)
```

Comment:
Running the validation with the update logic on estimation can potentially take a very long time. Probably better to check if all data up to date, if yes then pass validation and get estimate, else return fallback and update in the bg?

Reply:
The thought here was during normal runtime, the websocket updates should keep state up to date and `validate` will only take a handful of ms to run. Worst case, this would perform the duties of the websocket if we lost connection or something and fetch the necessary data from the node similarly to how we were previously getting gas prices from the node.